### DISCUSSION TITLE
Waiting for load-balancer to get an IP... loop (Probably caused by Rancher not supporting latest kube versions)

### DISCUSSION THREAD

**Discussion by Keyruu** [kube-hetzner/terraform-hcloud-kube-hetzner#789] [Q&A] [answered]
### Description

terraform apply
is stuck:
```
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization: Still creating... [1m30s elapsed]
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization: Still creating... [1m40s elapsed]
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
module.kube-hetzner.null_resource.kustomization (remote-exec): Waiting for load-balancer to get an IP...
```

The LoadBalancer service is created in Kubernetes and is healthy in Hetzner.

I saw another issue with the same issue. https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/539
I commented on there because I have the opposite situation my service name is `ngx-ingress-nginx-controller` instead of `nginx-ingress-nginx-controller`

### Kube.tf file

```terraform
locals {
  # Fill first and foremost your Hetzner API token, found in your project, Security, API Token, of type Read & Write.
  hcloud_token = var.hcloud_token

}

module "kube-hetzner" {
  providers = {
    hcloud = hcloud
  }
  hcloud_token = local.hcloud_token

  source = "kube-hetzner/kube-hetzner/hcloud"

  version = "2.1.5"

  ssh_public_key = file("/Users/user/.ssh/id_ed25519.pub")

  ssh_private_key = file("/Users/user/.ssh/id_ed25519")

  ssh_hcloud_key_label = "role=admin"

  network_region = "eu-central" # change to `us-east` if location is ash

  control_plane_nodepools = [
    {
      name        = "captain-nbg1",
      server_type = "cx21",
      location    = "nbg1",
      labels      = [],
      taints      = [],
      count       = 3
    },
  ]

  agent_nodepools = [
    {
      name        = "ship-small",
      server_type = "cx21",
      location    = "nbg1",
      labels      = [],
      taints      = [],
      count       = 0
    },
    {
      name        = "ship-large",
      server_type = "cx31",
      location    = "nbg1",
      labels      = [],
      taints      = [],
      count       = 8
    }
  ]

  load_balancer_type     = "lb11"
  load_balancer_location = "nbg1"

  etcd_s3_backup = {
    etcd-s3-access-key = var.aws_key
    etcd-s3-secret-key = var.aws_secret_key
    etcd-s3-bucket     = "k3s-etcd-snapshots"
    etcd-s3-region     = "eu-central-1"
  }


  ingress_controller = "nginx"

  cluster_name = "galaxy"

  extra_firewall_rules = [
    {
      description     = "To allow acces to resources via SSH"
      direction       = "out"
      protocol        = "tcp"
      port            = "22"
      source_ips      = [] # Won't be used for this rule
      destination_ips = ["0.0.0.0/0", "::/0"]
    }
  ]

  cni_plugin = "cilium"

  enable_cert_manager = true

  use_control_plane_lb = true

  lb_hostname = "cluster.io"

  enable_rancher = false

  rancher_hostname = "rancher.cluster.io"

  rancher_values = <<EOT
ingress:
  tls:
    source: "letsEncrypt"
hostname: "rancher.cluster.io"
letsEncrypt:
  email: admin@cluster.io
replicas: 3
bootstrapPassword: "password"
  EOT 

}

provider "hcloud" {
  token = local.hcloud_token
}

terraform {
  required_version = ">= 1.3.3"
  required_providers {
    hcloud = {
      source  = "hetznercloud/hcloud"
      version = ">= 1.38.2"
    }
  }
}

output "kubeconfig" {
  value     = module.kube-hetzner.kubeconfig
  sensitive = true
}
```

### Platform

Mac

**Comment by Keyruu** [comment]
I just had a look into the init.tf. It tries to search for the service in the nginx namespace but my service is located in kube-system.

**Comment by Keyruu** [comment]
I deployed everything with the terraform module.

**Comment by aDingil** [comment]
i just setup a new cluster, maybe hetzner is the culprit

**Comment by mysticaltech** [comment]
Thanks for confirming @aDingil!

@Keyruu Everything looks fine on my end too, just tried a test cluster with nginx and I get the LB just fine. Please note that there is an issue with your kube.tf file that could be indirectly causing this.

Rancher is not yet compatible with the latest version of Kubenetes, please check on their site (and read the explanation well around the rancher variables in kube.tf, there are also links in there). A safe bet would be to set `initial_k3s_channel` to v1.24 or v1.25.

Some  screenshots below from my test just now.

![ksnip_20230504-004304](https://user-images.githubusercontent.com/518555/236066732-7a261e30-4cf3-4863-8aaf-a1e358425acd.png)

![ksnip_20230504-004243](https://user-images.githubusercontent.com/518555/236066737-65e6ac74-57f0-4ab0-9254-99949e4cff6a.png)

**Reply by mysticaltech** [reply]
So @Keyruu just run destroy (see readme), set the a lower kube version as explained above and try again.

**Reply by Keyruu** [reply]
I can't take down the cluster. It's used in production...
And this isn't a new cluster. I created this a while back, maybe 6 months ago with version 1.6.5.
I just think this is a compatibility problem because I upgraded.
Back then the services were named like that: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/v1.6.5/locals.tf#LL304C6-L304C6

But now it's the other name: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/v2.1.5/locals.tf#L329

**Reply by mysticaltech** [reply]
Ok, yes the name changed between versions. So just destroy the old Traefik helmchartconfig, and apply again. List them with `kubectl get helmchartconfig`, you can check `kubectl get helmchart` if curious. More on this here https://docs.k3s.io/helm.

**Reply by Keyruu** [reply]
Will these be impacting the uptime of the cluster?

**Reply by Keyruu** [reply]
Will the HelmChart automatically be applied if I run terraform apply again?

**Reply by Keyruu** [reply, marked as answer]
Okay I fixed it with removing the old Helm Chart, which was called ngx and applying the templated yaml from terraform manually, so a new Ingress will be created. With this also comes a new LoadBalancer. This should probably be mentioned in the docs. I wasted a lot of hours trying to dig into this because the terraform module isn't really easily upgradable.

**Reply by Keyruu** [reply]
Thank you for the help though.

**Reply by mysticaltech** [reply]
No problem! FYI, applying the new Terraform would have worked too, it would actually be preferred. But glad you found a way!
